src/chains/ethereum/ethereum/src/transaction-pool.ts

```diff
@@ -130,14 +146,45 @@ export default class TransactionPool extends Emittery<{ drain: undefined }> {
     this.#priceBump = options.priceBump;
   }
   public readonly executables: Executables = {
-    inProgress: new Set(),
+    inProgress: new Map(),
     pending: new Map()
   };
   public readonly origins: Map<string, Heap<TypedTransaction>>;
-  readonly #accountPromises = new Map<
-    string,
-    Promise<{ balance: Quantity; nonce: Quantity }>
-  >();
+
+  /**
+   * Queues transactions to be inserted into the pool via `_prepareTransaction`
+   * such that unique-origin transactions are added to the pool concurrently and
+   * same-origin transactions are queued in series.
+   * @param transaction The transaction to be added to the transaction pool.
+   * @param secretKey The optional key to sign and hash the transaction.
+   * @returns Data that can be used to drain the queue.
+   */
+  public async prepareTransaction(
+    transaction: TypedTransaction,
+    secretKey?: Data
+  ) {
+    const origin = transaction.from.toString();
+    const originsQueue = this.#originsQueue;
+    let queueForOrigin: Semaphore;
+    // each unique origin gets its own semaphore rather than them all sharing
+    // one because we only need transactions from the same origin to be added in
+    // series
+    if (!(queueForOrigin = originsQueue.get(origin))) {
+      queueForOrigin = Semaphore(1);
+      originsQueue.set(origin, queueForOrigin);
+    }
+    // the semaphore will hold on `take` until previous uses have resolved
+    await new Promise(resolve => queueForOrigin.take(resolve));
+    try {
+      // check if transaction is in origins, if not add it
+      // if so, use semaphore to call prepareTransaction
+      // if not, just call prepare transaction
+      return await this._prepareTransaction(transaction, secretKey);
+    } finally {
+      // free up the semaphore once we finish
+      queueForOrigin.leave();
```

Comment:
Note: I think we could use the [semaphore's `available()` function](https://github.com/abrkn/semaphore.js/blob/67e14f8599563a37de1a8ca411d501b8aadb388b/lib/semaphore.js#L80) to see if there's any transaction waiting to be prepared in _this_ semaphore. If not, we could dispose of this semaphore. But then, if we get another transaction from that origin we have to make a whole new semaphore for it. 

So currently we don't drain the `originsQueue`. I'm interested in hearing y'all's thoughts on if we should drain it.

It's easy to imagine someone sending a transaction at regular intervals from the same origin, and this interval being slower than it takes to queue up the previous transaction. In that case, we'd be making a new semaphore for every transaction, which is why I decided against draining.

However, someone could also send a bunch of transactions, each from a unique origin. In that case, this `originsQueue` will just keep getting wider and wider without taking advantage of its purpose.